polygeist/tools/cgeist/Lib/CGExpr.cc

```diff
@@ -1939,117 +1814,51 @@ ValueCategory MLIRScanner::VisitCastExpr(CastExpr *E) {
-          Scalar);
-    }
-    if (auto LT = Scalar.getType().dyn_cast<mlir::LLVM::LLVMPointerType>()) {
-      mlir::Type MLIRTy = Glob.getTypes().getMLIRType(E->getType());
-      auto Val = Builder.create<mlir::LLVM::PtrToIntOp>(Loc, MLIRTy, Scalar);
-      return ValueCategory(Val, /*isReference*/ false);
-    }
-    LLVM_DEBUG({
-      Function.dump();
-      llvm::errs() << "Scalar: " << Scalar << "\n";
-      E->dump();
-    });
-    llvm_unreachable("unhandled ptrtoint cast");
-  } break;
-  case clang::CastKind::CK_IntegralToBoolean: {
-    auto Res = Visit(E->getSubExpr()).getValue(Builder);
-    auto PrevTy = Res.getType().cast<mlir::IntegerType>();
-    Res = Builder.create<arith::CmpIOp>(
-        Loc, arith::CmpIPredicate::ne, Res,
-        Builder.create<arith::ConstantIntOp>(Loc, 0, PrevTy));
-    auto PostTy =
-        Glob.getTypes().getMLIRType(E->getType()).cast<mlir::IntegerType>();
-    bool SignedType = true;
-    if (const auto *Bit = dyn_cast<clang::BuiltinType>(&*E->getType())) {
-      if (Bit->isUnsignedInteger())
-        SignedType = false;
-      if (Bit->isSignedInteger())
-        SignedType = true;
-    }
-    if (PostTy.getWidth() > 1) {
-      if (SignedType)
-        Res = Builder.create<arith::ExtSIOp>(Loc, PostTy, Res);
-      else
-        Res = Builder.create<arith::ExtUIOp>(Loc, PostTy, Res);
-    }
-    return ValueCategory(Res, /*isReference*/ false);
-  }
-  case clang::CastKind::CK_FloatingToBoolean: {
-    auto Res = Visit(E->getSubExpr()).getValue(Builder);
-    auto PrevTy = Res.getType().cast<mlir::FloatType>();
-    auto PostTy =
-        Glob.getTypes().getMLIRType(E->getType()).cast<mlir::IntegerType>();
-    bool SignedType = true;
-    if (const auto *Bit = dyn_cast<clang::BuiltinType>(&*E->getType())) {
-      if (Bit->isUnsignedInteger())
-        SignedType = false;
-      if (Bit->isSignedInteger())
-        SignedType = true;
-    }
-    auto Zero = Builder.create<arith::ConstantFloatOp>(
-        Loc, APFloat::getZero(PrevTy.getFloatSemantics()), PrevTy);
-    Res = Builder.create<arith::CmpFOp>(Loc, arith::CmpFPredicate::UNE, Res,
-                                        Zero);
-    if (1 < PostTy.getWidth()) {
-      if (SignedType)
-        Res = Builder.create<arith::ExtSIOp>(Loc, PostTy, Res);
-      else
-        Res = Builder.create<arith::ExtUIOp>(Loc, PostTy, Res);
-    }
-    return ValueCategory(Res, /*isReference*/ false);
-  }
+    const auto DestTy = E->getType();
+    assert(!DestTy->isBooleanType() && "bool should use PointerToBool");
+    auto PtrExpr = Visit(E->getSubExpr());
+
+    if (const auto MT = PtrExpr.val.getType().dyn_cast<MemRefType>())
+      PtrExpr = PtrExpr.MemRef2Ptr(Builder, Loc);
+
+    assert(PtrExpr.val.getType().isa<LLVM::LLVMPointerType>() &&
+           "Expecting LLVM pointer type");
+    const auto MLIRDestTy = Glob.getTypes().getMLIRType(DestTy);
+    return PtrExpr.PtrToInt(Builder, Loc, MLIRDestTy);
```

Comment:
Why we don't have something like `EmitPointerToIntCoversion`?

Reply:
True. Added!